### PR TITLE
Add Exoscale Import post-processor

### PR DESCRIPTION
**Pull request by falzm**
This change adds a new `exoscale-import` post-processor allowing users to create Private Templates on [Exoscale](https://www.exoscale.com/).

We can provide the Packer team with a free voucher for testing this code on the Exoscale platform if necessary.

Thanks!

### CODE REVIEW

File: post-processor/exoscale-import/post-processor.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+func init() {
+	egoscale.UserAgent = "Packer-Exoscale/" + version.FormattedVersion() + " " + egoscale.UserAgent
+}
+
+type PostProcessor struct {
+	config Config
+}
+
+func (p *PostProcessor) Configure(raws ...interface{}) error {
+	p.config.TemplateZone = defaultTemplateZone
```
Comment (SwampDragons) [inline post-processor/exoscale-import/post-processor.go]: If you overwrite the config values with the default in this way, then there's no point in having them as config values. Do you want them hardcoded, or do you want the user to be able to override them? If you want them hardcoded, just use the hardcoded var directly; if you want them user-overridable, you'll need to test for whether the user provided a value before assigning the default value.

File: website/source/docs/post-processors/exoscale-import.html.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+page_title: 'Exoscale Import - Post-Processors'
+sidebar_current: 'docs-post-processors-exoscale-import'
+---
+
+# Exoscale Import Post-Processor
+
+Type: `exoscale-import`
+
+The Packer Exoscale Import post-processor takes an image artifact from
+various builders and imports it to Exoscale.
```
Comment (SwampDragons) [inline website/source/docs/post-processors/exoscale-import.html.md]: Which specific artifact types can be imported? It'll be important for users to know what artifacts are allowed.  

You may want to add a test for this into the code itself. Example: 

https://github.com/hashicorp/packer/blob/master/post-processor/docker-import/post-processor.go#L48-L56

**Comment by falzm** [comment]
> A voucher would be helpful. My work email is megan at hashicorp dot com. I have a couple of comments inline that I'd like to get resolved. Also, can you also add yourself or someone else at Exoscale to the codeowners file?

I've sent an email with relevant information for you to get an account at Exoscale, and added myself and a backup teammate in the CODEOWNERS file.

**Comment by falzm** [comment]
(Sorry for the close/reopen, clicked on the wrong button... 😒)

File: post-processor/exoscale-import/post-processor.go (hunk 1 above)
Comment (falzm) [reply post-processor/exoscale-import/post-processor.go]: I'm not sure to understand your comment: I've pre-filled config values with default values *before* performing user configuration de-serialization into the config struct with `.Decode()`, so if they provide values in their config file they should overwrite the default values right? Did I miss something?

File: website/source/docs/post-processors/exoscale-import.html.md (hunk 2 above)
Comment (falzm) [reply website/source/docs/post-processors/exoscale-import.html.md]: Thanks, fixed.

File: post-processor/exoscale-import/post-processor.go (hunk 1 above)
Comment (SwampDragons) [reply post-processor/exoscale-import/post-processor.go]: Oh gosh, you're right of course. I didn't read far enough down.

File: website/source/docs/post-processors/exoscale-import.html.md (hunk 2 above)
Comment (SwampDragons) [inline website/source/docs/post-processors/exoscale-import.html.md]:
```suggestion
the QEMU, Artifice, or File builders and imports it to Exoscale.
```

**Comment by SwampDragons** [comment]
I tried to run a test of this import but I keep running into a "BucketRegionError"

Template:

```{
  "builders":
  [
	{
	  "type": "file",
	  "content": "Lorem ipsum dolor sit amet",
	  "target": "dummy_artifact"
	}
  ],
  "provisioners": [
      { "type": "shell",
        "inline": "echo hey"
      }
  ],
  "post-processors": [
  	{
	  "type": "exoscale-import",
	  "api_key": "keyvalue",
	  "api_secret": "secretvalue",
	  "template_zone": "at-vie-1",
	  "image_bucket": "packer-test-bucket",
	  "template_name": "myapp",
	  "template_username": "admin"

	}
  ]
}

```
object storage screenshot from exoscale console:
Screen Shot 2019-07-02 at 10 20 52 AM

Error message:

```
==> file: Running post-processor: exoscale-import
    file (exoscale-import): Uploading template image
dummy_artifact 26 B / 26 B [================================================================================================================================================] 100.00%
dummy_artifact 26 B / 26 B [=============================================================================================================================================] 100.00% 0s
2019/07/02 10:19:31 Deleting original artifact for build 'file'
2019/07/02 10:19:31 packer: 2019/07/02 10:19:31 Deleting dummy_artifact
2019/07/02 10:19:31 ui error: Build 'file' errored: 1 error(s) occurred:

* Post-processor failed: unable to upload image: BucketRegionError: incorrect region, the bucket is not in 'at-vie-1' region
	status code: 301, request id: , host id:
2019/07/02 10:19:31 machine readable: error-count []string{"1"}
2019/07/02 10:19:31 ui error:
==> Some builds didn't complete successfully and had errors:
Build 'file' errored: 1 error(s) occurred:

* Post-processor failed: unable to upload image: BucketRegionError: incorrect region, the bucket is not in 'at-vie-1' region
	status code: 301, request id: , host id:
2019/07/02 10:19:31 machine readable: file,error []string{"1 error(s) occurred:\n\n* Post-processor failed: unable to upload image: BucketRegionError: incorrect region, the bucket is not in 'at-vie-1' region\n\tstatus code: 301, request id: , host id: "}

2019/07/02 10:19:31 ui error: --> file: 1 error(s) occurred:

* Post-processor failed: unable to upload image: BucketRegionError: incorrect region, the bucket is not in 'at-vie-1' region
	status code: 301, request id: , host id:
==> Builds finished but no artifacts were created.
2019/07/02 10:19:31 waiting for all plugin processes to complete...
==> Some builds didn't complete successfully and had errors:
--> file: 1 error(s) occurred:

* Post-processor failed: unable to upload image: BucketRegionError: incorrect region, the bucket is not in 'at-vie-1' region
	status code: 301, request id: , host id:

==> Builds finished but no artifacts were created.
```

**Comment by SwampDragons** [comment]
I tried tweaking capitalization but that didn't work. Am I doing something wrong?

File: post-processor/exoscale-import/post-processor.go (hunk 1 above)
Comment (falzm) [reply post-processor/exoscale-import/post-processor.go]: OK good, I was afraid I've missed something big 😅

**Comment by falzm** [comment]
> I tried tweaking capitalization but that didn't work. Am I doing something wrong?

Right, I figured out what the problem is: you've created a bucket in the `at-vie-1` zone, however you're using the default `sos_endpoint` value which points to the `ch-gva-2` region (default value is `https://sos-ch-gva-2.exo.io`).

You must adjust your configuration:

```json
"sos_endpoint": "https://sos-at-vie-1.exo.io"
```

**Comment by falzm** [comment]
Also I've pushed I change to the post-processor: the `template_description` setting is actually mandatory, so please set a value for it (I've updated the PR accordingly).

**Comment by SwampDragons** [comment]
Do you think it's worth having a validation step to make sure that the region and endpoint match?

**Comment by SwampDragons** [comment]
Okay I was able to successfully build and upload a template! 🎉 

I'm going to merge this as-is but I would recommend doing two things to make the user experience better here. First, a validation check to make sure that the sos_endpoint matches the region, and second, some sanitization of regions to make capitalization not matter.  Those can be future PRs. good job on this :)

**Comment by falzm** [comment]
Thanks! I'll push upcoming PRs for the changes you've mentioned :)

**Pull request merged**